File: pet-allergy-scanner/SniffTest/Features/Nutrition/Services/WeightTrackingService.swift
//
//  WeightTrackingService.swift
//  SniffTest
//
//  Created by Steven Matos on 9/26/25.
//

import Foundation
import Combine

/**
 * Weight Tracking Service
 * 
 * Handles all weight tracking operations including:
 * - Weight recording and history management
 * - Goal setting and progress tracking
 * - Trend analysis and recommendations
 * - Data synchronization with backend
 * 
 * Follows SOLID principles with single responsibility for weight tracking
 * Implements DRY by reusing common data processing methods
 * Follows KISS by keeping the API simple and intuitive
 */
@MainActor
class WeightTrackingService: ObservableObject {
    static let shared = WeightTrackingService()
    
    @Published var weightHistory: [String: [WeightRecord]] = [:]
    @Published var weightGoals: [String: WeightGoal] = [:]
    @Published var currentWeights: [String: Double] = [:]
    @Published var recommendations: [String: [String]] = [:]
    @Published var isLoading = false
    @Published var error: Error?
    
    private let apiService: APIService
    private let unitService = WeightUnitPreferenceService.shared
    private let petService = PetService.shared
    private var cancellables = Set<AnyCancellable>()
    
    private init() {
        self.apiService = APIService.shared
    }
    
    // MARK: - Public API
    
    /**
     * Record a new weight measurement for a pet
     * - Parameter petId: The pet's ID
     * - Parameter weight: Weight in the user's selected unit
     * - Parameter notes: Optional notes about the measurement
     */
    func recordWeight(petId: String, weight: Double, notes: String? = nil) async throws {
        // Convert weight to kg for storage (backend expects kg)
        let weightInKg = unitService.convertToKg(weight)
        
        let weightRecord = WeightRecord(
            id: UUID().uuidString,
            petId: petId,
            weightKg: weightInKg,
            recordedAt: Date(),
            notes: notes,
            recordedByUserId: nil
        )
        
        // Add to local storage
        if weightHistory[petId] == nil {
            weightHistory[petId] = []
        }
        weightHistory[petId]?.insert(weightRecord, at: 0)
        currentWeights[petId] = weightInKg
        
        // Send to backend
        try await apiService.recordWeight(weightRecord)
        
        // Update pet's current weight in PetService
        await updatePetWeight(petId: petId, weightKg: weightInKg)
        
        // Generate recommendations
        await generateRecommendations(for: petId)
    }
    
    /**
     * Create or update a weight goal for a pet (one goal per pet)
     * - Parameter petId: The pet's ID
     * - Parameter goalType: Type of weight goal
     * - Parameter targetWeight: Target weight in the user's selected unit
     * - Parameter targetDate: Target date for achieving the goal
     * - Parameter notes: Optional notes about the goal
     */
    func upsertWeightGoal(
        petId: String,
        goalType: WeightGoalType,
        targetWeight: Double,
        targetDate: Date,
        notes: String? = nil
    ) async throws {
        // Convert target weight to kg for storage (backend expects kg)
        let targetWeightInKg = unitService.convertToKg(targetWeight)
        
        // Get current weight from pet or weight history - this will be the starting weight for the goal
        let currentWeightKg: Double?
        if let currentWeight = currentWeights[petId] {
            currentWeightKg = currentWeight
        } else if let latestRecord = weightHistory[petId]?.first {
            currentWeightKg = latestRecord.weightKg
        } else if let pet = petService.pets.first(where: { $0.id == petId }) {
            // Get weight from pet's profile as fallback
            currentWeightKg = pet.weightKg
        } else {
            // If no weight data available, we can't create a meaningful goal
            print("⚠️ Warning: No weight data available for pet \(petId), cannot create goal")
            currentWeightKg = nil
        }
        
        // Check if pet already has a goal
        let existingGoal = weightGoals[petId]
        let isUpdating = existingGoal != nil
        
        print("🎯 \(isUpdating ? "Updating" : "Creating") weight goal - Starting weight: \(currentWeightKg ?? 0), Target: \(targetWeightInKg), Goal type: \(goalType)")
        
        let weightGoal = WeightGoal(
            id: existingGoal?.id ?? UUID().uuidString, // Keep existing ID if updating
            petId: petId,
            goalType: goalType,
            targetWeightKg: targetWeightInKg,
            currentWeightKg: currentWeightKg,
            targetDate: targetDate,
            isActive: true,
            notes: notes,
            createdAt: existingGoal?.createdAt ?? Date(), // Keep original creation date if updating
            updatedAt: Date()
        )
        
        // Update local storage
        weightGoals[petId] = weightGoal
        print("✅ \(isUpdating ? "Updated" : "Created") weight goal locally: \(weightGoal.id)")
        
        // Send to backend
        do {
            try await apiService.createWeightGoal(weightGoal) // Backend now handles upsert
            print("✅ Successfully saved weight goal to backend: \(weightGoal.id)")
        } catch {
            print("❌ Failed to save weight goal to backend: \(error.localizedDescription)")
            // Don't throw error - keep the goal locally even if backend fails
            print("⚠️ Keeping goal locally despite backend failure")
        }
    }
    
    /**
     * Create a weight goal for a pet (deprecated - use upsertWeightGoal instead)
     * - Parameter petId: The pet's ID
     * - Parameter goalType: Type of weight goal
     * - Parameter targetWeight: Target weight in the user's selected unit
     * - Parameter targetDate: Target date for achieving the goal
     * - Parameter notes: Optional notes about the goal
     */
    func createWeightGoal(
        petId: String,
        goalType: WeightGoalType,
        targetWeight: Double,
        targetDate: Date,
        notes: String? = nil
    ) async throws {
        // Delegate to upsert method for consistency
        try await upsertWeightGoal(
            petId: petId,
            goalType: goalType,
            targetWeight: targetWeight,
            targetDate: targetDate,
            notes: notes
        )
    }
    
    /**
     * Update an existing weight goal for a pet (deprecated - use upsertWeightGoal instead)
     * - Parameter petId: The pet's ID
     * - Parameter goalType: Type of weight goal
     * - Parameter targetWeight: Target weight in the user's selected unit
     * - Parameter targetDate: Target date for achieving the goal
     * - Parameter notes: Optional notes about the goal
     */
    func updateWeightGoal(
        petId: String,
        goalType: WeightGoalType,
        targetWeight: Double,
        targetDate: Date,
        notes: String? = nil
    ) async throws {
        // Delegate to upsert method for consistency
        try await upsertWeightGoal(
            petId: petId,
            goalType: goalType,
            targetWeight: targetWeight,
            targetDate: targetDate,
            notes: notes
        )
    }
    
    /**
     * Load weight data for a pet
     * - Parameter petId: The pet's ID
     */
    func loadWeightData(for petId: String) async throws {
        isLoading = true
        error = nil
        
        do {
            // Load weight history
            let history = try await apiService.getWeightHistory(petId: petId)
            weightHistory[petId] = history
            
            // Load weight goals
            do {
                if let goal = try await apiService.getActiveWeightGoal(petId: petId) {
                    print("✅ Loaded weight goal from backend: \(goal.id)")
                    weightGoals[petId] = goal
                } else {
                    print("⚠️ No weight goal found in backend for pet: \(petId)")
                    // Clear local goal if no goal exists in backend
                    weightGoals[petId] = nil
                }
            } catch {
                // If goal loading fails, keep local goal but log the error
                print("❌ Failed to load weight goal from backend: \(error.localizedDescription)")
            }
            
            // Update current weight
            if let latestRecord = history.first {
                currentWeights[petId] = latestRecord.weightKg
            }
            
            // Generate recommendations
            await generateRecommendations(for: petId)
            
        } catch {
            self.error = error
            throw error
        }
        
        isLoading = false
    }
    
    /**
     * Get weight history for a pet
     * - Parameter petId: The pet's ID
     * - Returns: Array of weight records
     */
    func weightHistory(for petId: String) -> [WeightRecord] {
        return weightHistory[petId] ?? []
    }
    
    /**
     * Get current weight for a pet
     * - Parameter petId: The pet's ID
     * - Returns: Current weight or nil if not available
     */
    func currentWeight(for petId: String) -> Double? {
        return currentWeights[petId]
    }
    
    /**
     * Get active weight goal for a pet
     * - Parameter petId: The pet's ID
     * - Returns: Active weight goal or nil if not set
     */
    func activeWeightGoal(for petId: String) -> WeightGoal? {
        return weightGoals[petId]
    }
    
    /**
     * Get recommendations for a pet
     * - Parameter petId: The pet's ID
     * - Returns: Array of recommendation strings
     */
    func recommendations(for petId: String) -> [String] {
        return recommendations[petId] ?? []
    }
    
    /**
     * Analyze weight trend for a pet
     * - Parameter petId: The pet's ID
     * - Parameter daysBack: Number of days to analyze
     * - Returns: Weight trend analysis
     */
    func analyzeWeightTrend(for petId: String, daysBack: Int = 30) -> WeightTrendAnalysis {
        let history = weightHistory(for: petId)
        
        guard history.count >= 2 else {
            return WeightTrendAnalysis(
                trendDirection: .stable,
                weightChangeKg: 0.0,
                averageDailyChange: 0.0,
                trendStrength: .weak,
                daysAnalyzed: history.count,
                confidenceLevel: 0.0
            )
        }
        
        // Filter to specified days back
        let cutoffDate = Calendar.current.date(byAdding: .day, value: -daysBack, to: Date()) ?? Date()
        let recentHistory = history.filter { $0.recordedAt >= cutoffDate }
        
        guard recentHistory.count >= 2 else {
            return WeightTrendAnalysis(
                trendDirection: .stable,
                weightChangeKg: 0.0,
                averageDailyChange: 0.0,
                trendStrength: .weak,
                daysAnalyzed: recentHistory.count,
                confidenceLevel: 0.0
            )
        }
        
        // Calculate trend
        let currentWeight = recentHistory.first!.weightKg
        let oldWeight = recentHistory.last!.weightKg
        let weightChange = currentWeight - oldWeight
        
        // Calculate daily change
        let daysSpan = Calendar.current.dateComponents([.day], from: recentHistory.last!.recordedAt, to: recentHistory.first!.recordedAt).day ?? 1
        let dailyChange = weightChange / Double(daysSpan)
        
        // Determine trend direction
        let trendDirection: TrendDirection
        if weightChange > 0.5 {
            trendDirection = .increasing
        } else if weightChange < -0.5 {
            trendDirection = .decreasing
        } else {
            trendDirection = .stable
        }
        
        // Determine trend strength
        let absChange = abs(weightChange)
        let trendStrength: TrendStrength
        if absChange > 2.0 {
            trendStrength = .strong
        } else if absChange > 0.5 {
            trendStrength = .moderate
        } else {
            trendStrength = .weak
        }
        
        // Calculate confidence level
        let confidence = min(1.0, Double(recentHistory.count) / 14.0) // Max confidence at 2 weeks of data
        
        return WeightTrendAnalysis(
            trendDirection: trendDirection,
            weightChangeKg: round(weightChange * 100) / 100,
            averageDailyChange: round(dailyChange * 1000) / 1000,
            trendStrength: trendStrength,
            daysAnalyzed: recentHistory.count,
            confidenceLevel: confidence
        )
    }
    
    // MARK: - Private Methods
    
    /**
     * Update pet's current weight in PetService
     * - Parameter petId: Pet ID
     * - Parameter weightKg: New weight in kg
     */
    private func updatePetWeight(petId: String, weightKg: Double) async {
        // Find the pet in PetService and update its weight
        if petService.pets.firstIndex(where: { $0.id == petId }) != nil {
            let petUpdate = PetUpdate(
                name: nil,
                breed: nil,
                birthday: nil,
                weightKg: weightKg,
                activityLevel: nil,
                imageUrl: nil,
                knownSensitivities: nil,
                vetName: nil,
                vetPhone: nil
            )
            
            // Update the pet's weight through PetService
            await MainActor.run {
                petService.updatePet(id: petId, petUpdate: petUpdate)
            }
            
            // The PetService.updatePet method should handle the UI update
            // The pet's weight will be updated through the PetService
        }
    }
    
    private func generateRecommendations(for petId: String) async {
        let history = weightHistory(for: petId)
        let goal = activeWeightGoal(for: petId)
        let trend = analyzeWeightTrend(for: petId)
        
        var newRecommendations: [String] = []
        
        // Weight trend recommendations
        switch trend.trendDirection {
        case .increasing:
            if trend.trendStrength == .strong {
                newRecommendations.append("Rapid weight gain detected. Consider reducing portion sizes or increasing exercise.")
            } else {
                newRecommendations.append("Weight is trending upward. Monitor portion sizes.")
            }
        case .decreasing:
            if trend.trendStrength == .strong {
                newRecommendations.append("Significant weight loss detected. Ensure adequate nutrition and consult your veterinarian.")
            } else {
                newRecommendations.append("Weight is trending downward. Monitor food intake.")
            }
        case .stable:
            if let goal = goal {
                let currentWeight = currentWeights[petId] ?? 0
                let targetWeight = goal.targetWeightKg ?? 0
                
                if goal.goalType == .weightLoss && currentWeight > targetWeight {
                    newRecommendations.append("Continue current routine to reach your weight loss goal.")
                } else if goal.goalType == .weightGain && currentWeight < targetWeight {
                    newRecommendations.append("Consider increasing portion sizes to reach your weight gain goal.")
                } else {
                    newRecommendations.append("Great job maintaining your target weight!")
                }
            } else {
                newRecommendations.append("Weight is stable. Consider setting a weight goal for better tracking.")
            }
        }
        
        // Data quality recommendations
        if history.count < 7 {
            newRecommendations.append("Record weight more frequently for better trend analysis.")
        }
        
        // Goal recommendations
        if goal == nil {
            newRecommendations.append("Set a weight goal to track progress and get personalized recommendations.")
        }
        
        recommendations[petId] = newRecommendations
    }
}

// MARK: - Data Models

struct WeightRecord: Identifiable, Codable {
    let id: String
    let petId: String
    let weightKg: Double
    let recordedAt: Date
    let notes: String?
    let recordedByUserId: String?
    
    enum CodingKeys: String, CodingKey {
        case id, petId, weightKg, recordedAt, notes, recordedByUserId
    }
}

struct WeightGoal: Identifiable, Codable {
    let id: String
    let petId: String
    let goalType: WeightGoalType
    let targetWeightKg: Double?
    let currentWeightKg: Double?
    let targetDate: Date?
    let isActive: Bool
    let notes: String?
    let createdAt: Date
    let updatedAt: Date
    
    enum CodingKeys: String, CodingKey {
        case id, petId, goalType, targetWeightKg, currentWeightKg, targetDate, isActive, notes, createdAt, updatedAt
    }
}

enum WeightGoalType: String, CaseIterable, Codable {
    case weightLoss = "weight_loss"
    case weightGain = "weight_gain"
    case maintenance = "maintenance"
    case healthImprovement = "health_improvement"
}

struct WeightTrendAnalysis {
    let trendDirection: TrendDirection
    let weightChangeKg: Double
    let averageDailyChange: Double
    let trendStrength: TrendStrength
    let daysAnalyzed: Int
    let confidenceLevel: Double
}


enum TrendStrength {
    case weak
    case moderate
    case strong
}

// MARK: - API Request/Response Models

struct WeightRecordCreate: Codable {
    let pet_id: String
    let weight_kg: Double
    let recorded_at: Date
    let notes: String?
    let recorded_by_user_id: String?
    
    enum CodingKeys: String, CodingKey {
        case pet_id
        case weight_kg
        case recorded_at
        case notes
        case recorded_by_user_id
    }
}

struct WeightRecordResponse: Codable {
    let id: String
    let pet_id: String
    let weight_kg: Double
    let recorded_at: Date
    let notes: String?
    let recorded_by_user_id: String?
    
    enum CodingKeys: String, CodingKey {
        case id
        case pet_id
        case weight_kg
        case recorded_at
        case notes
        case recorded_by_user_id
    }
}

struct WeightGoalCreate: Codable {
    let pet_id: String
    let goal_type: String
    let targetWeightKg: Double?
    let currentWeightKg: Double?
    let targetDate: Date?
    let notes: String?
    
    enum CodingKeys: String, CodingKey {
        case pet_id
        case goal_type
        case targetWeightKg
        case currentWeightKg
        case targetDate
        case notes
    }
}

struct WeightGoalUpdate: Codable {
    let goal_type: String
    let targetWeightKg: Double?
    let targetDate: Date?
    let notes: String?
    
    enum CodingKeys: String, CodingKey {
        case goal_type
        case targetWeightKg
        case targetDate
        case notes
    }
}

struct WeightGoalResponse: Codable {
    let id: String
    let pet_id: String
    let goal_type: String
    let targetWeightKg: Double?
    let currentWeightKg: Double?
    let targetDate: Date?
    let isActive: Bool
    let notes: String?
    let createdAt: Date
    let updatedAt: Date
    
    enum CodingKeys: String, CodingKey {
        case id
        case pet_id
        case goal_type
        case targetWeightKg
        case currentWeightKg
        case targetDate
        case isActive
        case notes
        case createdAt
        case updatedAt
    }
}

// MARK: - API Service Extensions

extension APIService {
    /**
     * Record a new weight measurement for a pet
     * - Parameter weightRecord: Weight record data
     */
    func recordWeight(_ weightRecord: WeightRecord) async throws {
        let requestBody = WeightRecordCreate(
            pet_id: weightRecord.petId,
            weight_kg: weightRecord.weightKg,
            recorded_at: weightRecord.recordedAt,
            notes: weightRecord.notes,
            recorded_by_user_id: weightRecord.recordedByUserId
        )
        
        let _: WeightRecordResponse = try await post(
            endpoint: "/advanced-nutrition/weight/record",
            body: requestBody,
            responseType: WeightRecordResponse.self
        )
    }
    
    /**
     * Create a new weight goal for a pet
     * - Parameter weightGoal: Weight goal data
     */
    func createWeightGoal(_ weightGoal: WeightGoal) async throws {
        let requestBody = WeightGoalCreate(
            pet_id: weightGoal.petId,
            goal_type: weightGoal.goalType.rawValue,
            targetWeightKg: weightGoal.targetWeightKg,
            currentWeightKg: weightGoal.currentWeightKg,
            targetDate: weightGoal.targetDate,
            notes: weightGoal.notes
        )
        
        let _: WeightGoalResponse = try await post(
            endpoint: "/advanced-nutrition/weight/goals",
            body: requestBody,
            responseType: WeightGoalResponse.self
        )
    }
    
    /**
     * Update an existing weight goal for a pet
     * - Parameter weightGoal: Updated weight goal data
     */
    func updateWeightGoal(_ weightGoal: WeightGoal) async throws {
        let requestBody = WeightGoalUpdate(
            goal_type: weightGoal.goalType.rawValue,
            targetWeightKg: weightGoal.targetWeightKg,
            targetDate: weightGoal.targetDate,
            notes: weightGoal.notes
        )
        
        let _: WeightGoalResponse = try await post(
            endpoint: "/advanced-nutrition/weight/goals/\(weightGoal.id)",
            body: requestBody,
            responseType: WeightGoalResponse.self
        )
    }
    
    /**
     * Get weight history for a pet
     * - Parameter petId: Pet ID
     * - Returns: Array of weight records
     */
    func getWeightHistory(petId: String) async throws -> [WeightRecord] {
        let response: [WeightRecordResponse] = try await get(
            endpoint: "/advanced-nutrition/weight/history/\(petId)",
            responseType: [WeightRecordResponse].self
        )
        
        return response.map { response in
            WeightRecord(
                id: response.id,
                petId: response.pet_id,
                weightKg: response.weight_kg,
                recordedAt: response.recorded_at,
                notes: response.notes,
                recordedByUserId: response.recorded_by_user_id
            )
        }
    }
    
    /**
     * Get active weight goal for a pet
     * - Parameter petId: Pet ID
     * - Returns: Active weight goal or nil
     */
    func getActiveWeightGoal(petId: String) async throws -> WeightGoal? {
        let response: WeightGoalResponse? = try await get(
            endpoint: "/advanced-nutrition/weight/goals/\(petId)/active",
            responseType: WeightGoalResponse?.self
        )
        
        guard let response = response else { return nil }
        
        return WeightGoal(
            id: response.id,
            petId: response.pet_id,
            goalType: WeightGoalType(rawValue: response.goal_type) ?? .maintenance,
            targetWeightKg: response.targetWeightKg,
            currentWeightKg: response.currentWeightKg,
            targetDate: response.targetDate,
            isActive: response.isActive,
            notes: response.notes,
            createdAt: response.createdAt,
            updatedAt: response.updatedAt
        )
    }
}

File: server/app/services/weight_tracking_service.py
"""
Weight Tracking Service
Handles pet weight management, goal tracking, and trend analysis
"""

from typing import List, Optional, Dict, Any, Tuple
from datetime import datetime, date, timedelta
from decimal import Decimal
import statistics
from ..database import get_supabase_client
from ..utils.logging_config import get_logger
from ..models.advanced_nutrition import (
    PetWeightRecordCreate, PetWeightRecordResponse,
    PetWeightGoalCreate, PetWeightGoalResponse,
    WeightTrendAnalysis, TrendDirection, TrendStrength,
    WeightManagementDashboard
)

logger = get_logger(__name__)


class WeightTrackingService:
    """
    Service for managing pet weight tracking and analysis
    
    Follows SOLID principles with single responsibility for weight management
    Implements DRY by reusing common calculation methods
    Follows KISS by keeping methods focused and simple
    """
    
    def __init__(self):
        self.supabase = get_supabase_client()
    
    async def record_weight(
        self, 
        weight_record: PetWeightRecordCreate,
        user_id: str
    ) -> PetWeightRecordResponse:
        """
        Record a new weight measurement for a pet
        
        Note: Pet ownership should be verified by the calling router/endpoint
        before calling this method. This method assumes the pet exists and
        the user has access.
        
        Args:
            weight_record: Weight record data
            user_id: ID of the user recording the weight (used for logging/audit)
            
        Returns:
            Created weight record
            
        Raises:
            ValueError: If weight recording fails
        """
        # Insert weight record using centralized service
        # Note: Pet ownership is verified by the router before calling this method
        weight_data = {
            "pet_id": weight_record.pet_id,
            "weight_kg": float(weight_record.weight_kg),
            "recorded_at": weight_record.recorded_at.isoformat(),
            "notes": weight_record.notes,
            "recorded_by_user_id": user_id
        }
        
        db_service = DatabaseOperationService(self.supabase)
        result = db_service.insert_with_timestamps("pet_weight_records", weight_data, include_created_at=False)
        
        # Update nutritional trends for this date
        await self._update_nutritional_trends(weight_record.pet_id, weight_record.recorded_at.date())
        
        return PetWeightRecordResponse(**result)
    
    async def get_weight_history(
        self, 
        pet_id: str, 
        user_id: str,
        days_back: int = 30
    ) -> List[PetWeightRecordResponse]:
        """
        Get weight history for a pet
        
        Note: Pet ownership should be verified by the calling router/endpoint
        before calling this method. This method assumes the pet exists and
        the user has access.
        
        Args:
            pet_id: Pet ID
            user_id: User ID for authorization (used for logging/audit)
            days_back: Number of days to look back
            
        Returns:
            List of weight records (empty list if no records exist)
        """
        # Get weight records
        # Note: Pet ownership is verified by the router before calling this method
        start_date = DateTimeService.now() - timedelta(days=days_back)
        
        response = self.supabase.table("pet_weight_records")\
            .select("*")\
            .eq("pet_id", pet_id)\
            .gte("recorded_at", start_date.isoformat())\
            .order("recorded_at", desc=True)\
            .execute()
        
        # Return empty list if no records exist (200 status with empty data)
        return [PetWeightRecordResponse(**record) for record in response.data]
    
    async def upsert_weight_goal(
        self, 
        goal: PetWeightGoalCreate,
        user_id: str
    ) -> PetWeightGoalResponse:
        """
        Create or update a weight goal for a pet (one goal per pet)
        
        Note: Pet ownership should be verified by the calling router/endpoint
        before calling this method. This method assumes the pet exists and
        the user has access.
        
        Args:
            goal: Weight goal data
            user_id: User ID for authorization (used for logging/audit)
            
        Returns:
            Created or updated weight goal
        """
        # Check if pet already has a weight goal
        # Note: Pet ownership is verified by the router before calling this method
        existing_goal_response = self.supabase.table("pet_weight_goals")\
            .select("*")\
            .eq("pet_id", goal.pet_id)\
            .execute()
        
        goal_data = {
            "pet_id": goal.pet_id,
            "goal_type": goal.goal_type.value,
            "target_weight_kg": float(goal.target_weight_kg) if goal.target_weight_kg else None,
            "current_weight_kg": float(goal.current_weight_kg) if goal.current_weight_kg else None,
            "target_date": goal.target_date.isoformat() if goal.target_date else None,
            "is_active": goal.is_active,
            "notes": goal.notes
        }
        
        if existing_goal_response.data:
            # Update existing goal
            goal_id = existing_goal_response.data[0]["id"]
            response = self.supabase.table("pet_weight_goals")\
                .update(goal_data)\
                .eq("id", goal_id)\
                .execute()
            
            if not response.data:
                raise ValueError("Failed to update weight goal")
            
            return PetWeightGoalResponse(**response.data[0])
        else:
            # Create new goal
            db_service = DatabaseOperationService(self.supabase)
            result = db_service.insert_with_timestamps("pet_weight_goals", goal_data)
            
            return PetWeightGoalResponse(**result)
    
    async def create_weight_goal(
        self, 
        goal: PetWeightGoalCreate,
        user_id: str
    ) -> PetWeightGoalResponse:
        """
        Create a weight goal for a pet (deprecated - use upsert_weight_goal instead)
        
        Args:
            goal: Weight goal data
            user_id: User ID for authorization
            
        Returns:
            Created weight goal
        """
        # Delegate to upsert method for consistency
        return await self.upsert_weight_goal(goal, user_id)
    
    async def get_active_weight_goal(
        self, 
        pet_id: str, 
        user_id: str
    ) -> Optional[PetWeightGoalResponse]:
        """
        Get active weight goal for a pet
        
        Note: Pet ownership should be verified by the calling router/endpoint
        before calling this method. This method assumes the pet exists and
        the user has access.
        
        Args:
            pet_id: Pet ID
            user_id: User ID for authorization (used for logging/audit)
            
        Returns:
            Active weight goal or None (returns None if no goal exists)
        """
        # Get active goal
        # Note: Pet ownership is verified by the router before calling this method
        response = self.supabase.table("pet_weight_goals")\
            .select("*")\
            .eq("pet_id", pet_id)\
            .eq("is_active", True)\
            .execute()
        
        if not response.data:
            return None
        
        return PetWeightGoalResponse(**response.data[0])
    
    async def analyze_weight_trend(
        self, 
        pet_id: str, 
        user_id: str,
        days_back: int = 30
    ) -> WeightTrendAnalysis:
        """
        Analyze weight trend for a pet
        
        Args:
            pet_id: Pet ID
            user_id: User ID for authorization
            days_back: Number of days to analyze
            
        Returns:
            Weight trend analysis
        """
        # Get weight history
        weight_records = await self.get_weight_history(pet_id, user_id, days_back)
        
        if len(weight_records) < 2:
            return WeightTrendAnalysis(
                trend_direction=TrendDirection.STABLE,
                weight_change_kg=0.0,
                average_daily_change=0.0,
                trend_strength=TrendStrength.WEAK,
                days_analyzed=len(weight_records),
                confidence_level=0.0
            )
        
        # Sort by date
        sorted_records = sorted(weight_records, key=lambda x: x.recorded_at)
        
        # Calculate trend
        current_weight = sorted_records[-1].weight_kg
        old_weight = sorted_records[0].weight_kg
        weight_change = current_weight - old_weight
        
        # Calculate daily change
        days_span = (sorted_records[-1].recorded_at - sorted_records[0].recorded_at).days
        daily_change = weight_change / days_span if days_span > 0 else 0
        
        # Determine trend direction
        if weight_change > 0.5:
            trend_direction = TrendDirection.INCREASING
        elif weight_change < -0.5:
            trend_direction = TrendDirection.DECREASING
        else:
            trend_direction = TrendDirection.STABLE
        
        # Determine trend strength
        abs_change = abs(weight_change)
        if abs_change > 2.0:
            trend_strength = TrendStrength.STRONG
        elif abs_change > 0.5:
            trend_strength = TrendStrength.MODERATE
        else:
            trend_strength = TrendStrength.WEAK
        
        # Calculate confidence level based on data points and consistency
        confidence = min(1.0, len(weight_records) / 14)  # Max confidence at 2 weeks of data
        
        return WeightTrendAnalysis(
            trend_direction=trend_direction,
            weight_change_kg=round(weight_change, 2),
            average_daily_change=round(daily_change, 3),
            trend_strength=trend_strength,
            days_analyzed=len(weight_records),
            confidence_level=confidence
        )
    
    async def get_weight_management_dashboard(
        self, 
        pet_id: str, 
        user_id: str
    ) -> WeightManagementDashboard:
        """
        Get comprehensive weight management dashboard data
        
        Args:
            pet_id: Pet ID
            user_id: User ID for authorization
            
        Returns:
            Weight management dashboard data
        """
        # Get current weight (most recent)
        weight_records = await self.get_weight_history(pet_id, user_id, 1)
        current_weight = weight_records[0].weight_kg if weight_records else None
        
        # Get active weight goal
        weight_goal = await self.get_active_weight_goal(pet_id, user_id)
        target_weight = weight_goal.target_weight_kg if weight_goal else None
        
        # Get weight trend
        weight_trend = await self.analyze_weight_trend(pet_id, user_id, 30)
        
        # Get weekly progress (last 4 weeks)
        weekly_progress = await self._get_weekly_progress(pet_id, user_id, 4)
        
        # Get recommendations (would be generated by recommendation service)
        recommendations = []  # TODO: Integrate with recommendation service
        
        return WeightManagementDashboard(
            pet_id=pet_id,
            current_weight=current_weight,
            target_weight=target_weight,
            weight_goal=weight_goal,
            recent_trend=weight_trend,
            weekly_progress=weekly_progress,
            recommendations=recommendations
        )
    
    async def _deactivate_existing_goals(self, pet_id: str) -> None:
        """Deactivate existing active goals for a pet"""
        self.supabase.table("pet_weight_goals")\
            .update({"is_active": False})\
            .eq("pet_id", pet_id)\
            .eq("is_active", True)\
            .execute()
    
    async def _update_nutritional_trends(self, pet_id: str, trend_date: date) -> None:
        """Update nutritional trends for a specific date"""
        # This would call the database function to update trends
        # For now, we'll implement a simple version
        try:
            self.supabase.rpc("update_nutritional_trends", {
                "pet_uuid": pet_id,
                "trend_date": trend_date.isoformat()
            }).execute()
        except Exception as e:
            # Log error but don't fail the weight recording
            logger.warning(f"Failed to update nutritional trends: {e}")
    
    async def _get_weekly_progress(
        self, 
        pet_id: str, 
        user_id: str, 
        weeks: int
    ) -> List[Dict[str, Any]]:
        """Get weekly weight progress data"""
        weekly_data = []
        
        for week_offset in range(weeks):
            now = DateTimeService.now()
            week_start = now - timedelta(weeks=week_offset + 1)
            week_end = week_start + timedelta(days=6)
            
            # Get weight records for this week
            week_records = await self.get_weight_history(
                pet_id, 
                user_id, 
                (week_end - now).days + 7
            )
            
            # Filter to this week
            week_records = [
                r for r in week_records 
                if week_start.date() <= r.recorded_at.date() <= week_end.date()
            ]
            
            if week_records:
                start_weight = week_records[0].weight_kg
                end_weight = week_records[-1].weight_kg
                weight_change = end_weight - start_weight
                
                weekly_data.append({
                    "week_start": week_start.date().isoformat(),
                    "week_end": week_end.date().isoformat(),
                    "start_weight": start_weight,
                    "end_weight": end_weight,
                    "weight_change": weight_change,
                    "measurements": len(week_records)
                })
        
        return weekly_data
